File: src/day19/main.rs
#![allow(soft_unstable)]
use std::collections::VecDeque;

use itertools::{izip, Itertools};
use regex::Regex;
use shared::puzzle_input;

const ALLOW_ALL: [bool; 4] = [true, true, true, true];

type Blueprint = [[u8; 3]; 4];

type Counts = [u8; 4];

#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone)]
struct State {
    bot_counts: Counts,
    resources: Counts,
}

fn can_afford(costs: &[u8; 3], resources: &Counts) -> bool {
    costs.iter().zip(resources.iter()).all(|(a, b)| a <= b)
}

fn spend(mut resources: Counts, costs: &[u8; 3]) -> Counts {
    for i in 0..3 {
        resources[i] -= costs[i];
    }
    resources
}

fn gather(mut resources: Counts, bots: &Counts) -> Counts {
    for i in 0..4 {
        resources[i] += bots[i];
    }
    resources
}

fn play(bp: &Blueprint, minutes: u8) -> u8 {
    // PRUNING - changed from DFS to BFS so that we can check when things are falling behind with just 1 number
    let mut queue = VecDeque::<_>::new();
    queue.push_back((
        State {
            bot_counts: [1, 0, 0, 0],
            resources: Counts::default(),
        },
        ALLOW_ALL,
        1u8,
    ));
    let caps = izip!(bp[0].iter(), bp[1].iter(), bp[2].iter(), bp[3].iter())
        .map(|(a, b, c, d)| *[a, b, c, d].iter().max().unwrap())
        .collect_vec();
    let mut best = 0;
    'next: while let Some((state, allow, time)) = queue.pop_front() {
        best = best.max(state.resources[3]);

        if time > minutes {
            continue;
        }

        // PRUNING - falling behind the rest of the BFS
        if state.resources[3] + state.bot_counts[3] + 1 < best {
            continue;
        }

        // PRUNING - if we skipped building something we could, no point in considering building it next time
        let allow_after_skip = [0, 1, 2, 3].map(|i| !can_afford(&bp[i], &state.resources));

        for bot_type in (0..4).rev() {
            let full = (bot_type != 3) && (state.bot_counts[bot_type] >= *caps[bot_type]);
            if !full && allow[bot_type] && !allow_after_skip[bot_type] {
                let resources = spend(state.resources, &bp[bot_type]);
                let resources = gather(resources, &state.bot_counts);
                let mut bot_counts = state.bot_counts;
                bot_counts[bot_type] += 1;
                queue.push_back((
                    State {
                        bot_counts,
                        resources,
                    },
                    ALLOW_ALL,
                    time + 1,
                ));
                // PRUNING - if we built a geode bot, don't consider other bots or skipping
                if bot_type == 3 {
                    continue 'next;
                }
            }
        }

        let resources = gather(state.resources, &state.bot_counts);
        queue.push_back((State { resources, ..state }, allow_after_skip, time + 1));
    }
    best
}

pub fn main() {
    use std::time::Instant;
    let now = Instant::now();
    let (p1, p2) = {
        let sep = Regex::new(r"[^0-9]+").unwrap();
        let input = puzzle_input!();
        let bps = input
            .split('\n')
            .map(|line| {
                let v = sep
                    .split(line)
                    .filter(|s| !s.is_empty())
                    .map(|s| s.parse::<u8>().unwrap())
                    .collect_vec();
                [[v[1], 0, 0], [v[2], 0, 0], [v[3], v[4], 0], [v[5], 0, v[6]]]
            })
            .collect_vec();
        let p1 = bps
            .iter()
            .map(|bp| play(bp, 24) as usize)
            .enumerate()
            .map(|(i, v)| (i + 1) * v)
            .sum::<usize>();
        let p2 = bps
            .iter()
            .take(3)
            .map(|bp| play(bp, 32) as usize)
            .product::<usize>();
        (p1, p2)
    };
    let elapsed = now.elapsed();
    eprintln!("Elapsed: {:.2?}", elapsed);

    // 1395, 2700
    println!("part 1: {}\npart 2: {}", p1, p2);
}

File: src/day19/main.exs
defmodule Day19 do
  @moduledoc "AoC2022 Day19 functions"
  alias ElixirDS.Deque

  def play(bp, minutes) do
    deq = Deque.from_list([{[1, 0, 0, 0], [0, 0, 0, 0], [true, true, true, true], 1}])
    caps = Enum.zip(bp) |> Enum.map(fn {a, b, c, d} -> Enum.max([a, b, c, d]) end)
    play(bp, minutes, caps, deq, 0)
  end

  defp play(bp, minutes, caps, deq, best) do
    # breadth-first so that we can easily tell when it's falling behind
    case Deque.pop_front(deq) do
      :error ->
        best

      {{bots, resources, allow, time}, deq} ->
        [_, _, _, geodes] = resources
        best = max(best, geodes)

        if time > minutes do
          play(bp, minutes, caps, deq, best)
        else
          [_, _, _, bots_geodes] = bots
          # PRUNING - falling behind the rest of the BFS
          if geodes + bots_geodes + 1 < best do
            play(bp, minutes, caps, deq, best)
          else
            can_afford = Enum.map(bp, fn cost -> can_afford(cost, resources) end)

            # PRUNING - if we skipped building something we could, no point in considering building it next time
            allow_after_skip = Enum.map(can_afford, &(not &1))

            {stop, deq} =
              Enum.zip([caps, bots, allow, bp, can_afford])
              |> Enum.with_index()
              |> Enum.reverse()
              |> Enum.flat_map(fn {{cap, bot_count, allow, cost, can_afford}, i} ->
                full = i !== 3 and bot_count >= cap

                if !full and allow and can_afford do
                  [{i, build(bots, cost, resources, time, i)}]
                else
                  []
                end
              end)
              # PRUNING - stop here is so that if we built a geode bot, don't consider other bots or skipping
              |> Enum.reduce({false, deq}, fn {i, item}, {stop, deq} ->
                if stop do
                  {true, deq}
                else
                  {i == 3, Deque.push_back(deq, item)}
                end
              end)

            deq =
              if stop do
                deq
              else
                resources = Enum.zip(resources, bots) |> Enum.map(fn {r, b} -> r + b end)
                Deque.push_back(deq, {bots, resources, allow_after_skip, time + 1})
              end

            play(bp, minutes, caps, deq, best)
          end
        end
    end
  end

  defp build(bot_counts, cost, resources, time, i) do
    resources = Enum.zip([resources, cost, bot_counts]) |> Enum.map(fn {r, c, b} -> r - c + b end)
    bot_counts = List.update_at(bot_counts, i, &(&1 + 1))
    {bot_counts, resources, [true, true, true, true], time + 1}
  end

  defp can_afford(costs, resources) do
    Enum.zip(costs, resources)
    |> Enum.all?(fn {a, b} -> a <= b end)
  end
end

bps =
  File.read!("puzzle_input/day19")
  |> String.trim_trailing()
  |> String.split(~r/[^0-9]+/, trim: True)
  |> Enum.filter(fn s -> s !== "" end)
  |> Enum.map(&String.to_integer/1)
  |> Enum.chunk_every(7)
  |> Enum.map(fn [_, x1, x2, x3, x4, x5, x6] ->
    [[x1, 0, 0, 0], [x2, 0, 0, 0], [x3, x4, 0, 0], [x5, 0, x6, 0]]
  end)

p1 =
  Enum.map(bps, fn bp ->
    Day19.play(bp, 24)
  end)
  |> Enum.with_index(1)
  |> Enum.map(fn {x, i} -> x * i end)
  |> Enum.sum()

p2 = Enum.take(bps, 3) |> Enum.map(fn bp -> Day19.play(bp, 32) end) |> Enum.product()

# 1395, 2700
IO.puts("part 1: #{p1}\npart 2: #{p2}")
